threestudio/models/networks.py

```diff
@@ -51,6 +55,90 @@ def update_step(self, epoch, global_step, on_load_weights=False):
                 f"Update mask: {global_step}/{self.n_masking_step} {self.mask}"
             )
 
+class KPlanesFeaturePlane(torch.nn.Module):
+    def __init__(
+        self,
+        feature_dim: int = 8,
+        resolution: Tuple[int, int] = (128, 128),
+        init: Callable = torch.nn.init.uniform_
+    ):
+        super().__init__()
+        self.feature_dim = feature_dim
+        with torch.cuda.device(get_rank()):
+            self.plane = torch.nn.Parameter(torch.empty(1, feature_dim, *resolution))
+        init(self.plane)
+
+    def forward(self, x: Float[torch.Tensor, "*N 2"]) -> Float[torch.Tensor, "*N C"]:   
+        # x is in the range 0 to 1, transform to -1 to 1. 
+        x = (x * 2) - 1
+        new_shape = [*x.size()[:-1],self.feature_dim]
+        output = torch.nn.functional.grid_sample(
+            self.plane,
+            x.view(1,-1,1,2),
+            align_corners=True 
+        ).squeeze().transpose(0,-1).contiguous()
+        return output.view(new_shape)
+
+    def loss_tv(self) -> torch.Tensor:
+        tv_x = torch.nn.functional.mse_loss(self.plane[:, :, 1:, :], self.plane[:, :, :-1, :])
+        tv_y = torch.nn.functional.mse_loss(self.plane[:, :, :, 1:], self.plane[:, :, :, :-1])
+        return tv_x + tv_y
+
+    def loss_l1(self) -> torch.Tensor:
+        return torch.mean(torch.abs(self.plane))
+
+class KPlanesFeatureField(torch.nn.Module):
+    def __init__(self, n_input_dims : int, config):        
+        super().__init__()
+
+        hierarchical_list = []
+        for resolution in config["resolutions"]: # e.g resolutions=[32,64,128]
+            plane_list = [KPlanesFeaturePlane(config["n_feature_count"], resolution=(resolution,resolution))] * config["n_planes"]
```

Comment:
Nope, `config["n_planes"]` should be `math.comb(config["k"], 2)`.
Or, `n_planes` should be 6 for 4D